File: employeeschedule.py
import random

def collect_employee_data():
    while True:
        try:
            num_employees = int(input("Enter the number of employees: "))
            if num_employees <= 0:
                print("Please enter a valid positive number.")
                continue
            break
        except ValueError:
            print("Invalid input! Please enter a number.")

    employees = {}
    for _ in range(num_employees):
        name = input("Enter employee name: ").strip()
        employees[name] = {
            "assigned_shifts": {},  # Store assigned shifts per day
            "assigned_days": 0      # Track number of days worked
        }

    return employees


def assign_shifts(employees):
    days = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
    shifts = ["Morning", "Afternoon", "Evening"]
    schedule = {day: {shift: [] for shift in shifts} for day in days}

    for day in days:
        for shift in shifts:
            for employee in employees:
                if employees[employee]["assigned_days"] < 5 and len(schedule[day][shift]) < 2:
                    schedule[day][shift].append(employee)
                    employees[employee]["assigned_days"] += 1  # Track workdays

    return schedule


def display_schedule(schedule):
    print("\nFinal Employee Schedule:")
    for day, shifts in schedule.items():
        print(f"\n{day}:")
        for shift, employees in shifts.items():
            print(f"  {shift.capitalize()}: {', '.join(employees) if employees else 'No employees assigned'}")

if __name__ == "__main__":
    employees = collect_employee_data()
    schedule = assign_shifts(employees)
    display_schedule(schedule)

File: employeeschedule.cpp
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>
#include <ctime>
#include <cstdlib>

using namespace std;

struct Employee {
    string name;
    vector<string> preferred_shifts;
    int assigned_days;
};

const vector<string> days = {"Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"};
const vector<string> shifts = {"Morning", "Afternoon", "Evening"};

map<string, map<string, vector<string>>> schedule;

void assignShifts(vector<Employee>& employees) {
    srand(time(0)); // Seed for randomness
    
    for (const string& day : days) {
        for (const string& shift : shifts) {
            vector<string> assigned;
            for (Employee& employee : employees) {
                if (employee.assigned_days < 5 && find(employee.preferred_shifts.begin(), employee.preferred_shifts.end(), shift) != employee.preferred_shifts.end()) {
                    assigned.push_back(employee.name);
                    employee.assigned_days++;
                }
                if (assigned.size() == 2) break;
            }
            while (assigned.size() < 2) {
                for (Employee& employee : employees) {
                    if (employee.assigned_days < 5 && find(assigned.begin(), assigned.end(), employee.name) == assigned.end()) {
                        assigned.push_back(employee.name);
                        employee.assigned_days++;
                        if (assigned.size() == 2) break;
                    }
                }
            }
            schedule[day][shift] = assigned;
        }
    }
}

void printSchedule() {
    cout << "\nFinal Employee Schedule:" << endl;
    for (const string& day : days) {
        cout << "\n" << day << ":\n";
        for (const string& shift : shifts) {
            cout << shift << ": ";
            for (const string& employee : schedule[day][shift]) {
                cout << employee << " ";
            }
            cout << endl;
        }
    }
}

int main() {
    int numEmployees;
    cout << "Enter the number of employees: ";
    cin >> numEmployees;
    cin.ignore();
    
    vector<Employee> employees(numEmployees);
    for (int i = 0; i < numEmployees; ++i) {
        cout << "Enter employee name: ";
        getline(cin, employees[i].name);
        employees[i].assigned_days = 0;
        cout << "Enter preferred shifts (Morning, Afternoon, Evening) separated by space: ";
        string shift;
        getline(cin, shift);
        size_t pos = 0;
        while ((pos = shift.find(' ')) != string::npos) {
            employees[i].preferred_shifts.push_back(shift.substr(0, pos));
            shift.erase(0, pos + 1);
        }
        employees[i].preferred_shifts.push_back(shift);
    }
    
    assignShifts(employees);
    printSchedule();
    
    return 0;
}
